src/odemis/util/test/registration_test.py

```diff
@@ -0,0 +1,404 @@
+                tform = AffineTransform(matrix)
+                self.assertGreater(numpy.linalg.det(matrix), 0)
+                self.assertGreater(tform.rotation, -0.25 * numpy.pi)
+                self.assertLess(tform.rotation, 0.25 * numpy.pi)
+
+    def test_angle_difference(self):
+        """
+        When provided with an input transformation with positive determinant
+        (i.e. no reflection), the resulting change in rotation should be a
+        multiple of 90°.
+
+        """
+        for i in range(50):
+            matrix = self._rng.standard_normal((2, 2))
+            if numpy.linalg.det(matrix) < 0:
+                matrix = numpy.fliplr(matrix)
+            with self.subTest(matrix=matrix, i=i):
+                tform1 = AffineTransform(matrix)
+                tform2 = AffineTransform(_canonical_matrix_form(matrix))
+                delta = tform2.rotation - tform1.rotation
+                n = int(round(delta / (0.5 * numpy.pi)))
+                self.assertAlmostEqual(delta - n * 0.5 * numpy.pi, 0)
+
+
+class EstimateGridOrientationTest(unittest.TestCase):
+    """Unittest for `estimate_grid_orientation()`."""
+
+    def setUp(self):
+        """Ensure reproducible tests."""
+        self._rng = check_random_state(12345)
+
+    def test_full_grid(self):
+        """
+        `estimate_grid_orientation()` should return the expected
+        AffineTransform for a full grid.
+
+        """
+        for shape in [(5, 5), (8, 8)]:
+            for i in range(50):
+                with self.subTest(shape=shape, i=i):
+                    n, m = shape
+                    shuffle = self._rng.permutation(n * m)
+                    tform = random_affine_transform(self._rng)
+                    xy = tform.apply(unit_gridpoints(shape, mode="xy")[shuffle])
+                    out = estimate_grid_orientation(xy, shape, AffineTransform)
+                    for name in ("matrix", "translation"):
+                        numpy.testing.assert_array_almost_equal(
+                            getattr(tform, name), getattr(out, name)
+                        )
+
+    def test_incomplete_grid(self):
+        """
+        `estimate_grid_orientation()` should return the expected
+        AffineTransform for a grid with one point missing.
+
+        """
+        for shape in [(5, 5), (8, 8)]:
+            for i in range(50):
+                with self.subTest(shape=shape, i=i):
+                    n, m = shape
+                    shuffle = self._rng.permutation(n * m)
+                    tform = random_affine_transform(self._rng)
+                    xy = tform.apply(unit_gridpoints(shape, mode="xy")[shuffle])[:-1]
+                    out = estimate_grid_orientation(xy, shape, AffineTransform)
+                    for name in ("matrix", "translation"):
+                        numpy.testing.assert_array_almost_equal(
+                            getattr(tform, name), getattr(out, name)
+                        )
```

Comment:
same here, this is also hard to read, I would rewrite it to:

```
numpy.testing.assert_array_almost_equal(tform.matrix, out.matrix)
numpy.testing.assert_array_almost_equal(tform.translation, out.translation)
```

Reply:
done